### PR TITLE
restore random states after export_model

### PR DESCRIPTION
**Pull request by mindest**
**Description**: Found that random states may be changed after `export_model`. This change restores the random states after the export.

**Motivation and Context**
- Why is this change required? What problem does it solve?
On a 1P training model, the training data files are shuffled before each epoch using `random.shuffle()`. But we found that after exporting ONNX graph, the random state is randomly changed and this makes random states across different ranks different, leading to validation loss.
Further investigation of this case found that the random state is changed during **symbolic shape inference** when processing a `Slice` node, which includes a sympy function `sympy.Min()`. In fact, some other sympy functions can also change the random state, which is beyond our control. Here we record a few random states of interest (python random, numpy random, torch random, and torch.cuda random) and restore them after exporting to avoid future issues.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 1]
```diff
@@ -228,7 +228,9 @@ def forward(self, *inputs, **kwargs):
             build_gradient_graph = False
             if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_BUILD_GRADIENT) is False or \
                     not self._onnx_models.exported_model:
+                random_states = _utils.get_random_states()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:231]: Do we know why `self._export_model(...)` changes the random state? Is this introduced by the exporter or some code inside `ortmodule`?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:231]: Probably because the models is executed once on that seed during export. After model is executed, the random number generator will run as it would for the second time

That is why we do a deep copy of the model before exporting, to make sure any randomness inside the model is not changed during export. 

However, I am not sure whether making ORTModule aware of external random genrators is a good idea. There can be many out there, including custom ones that ORTModule will never handle

It is probably a better idea to document such limitation that states that during first step random generators can be modified and it is user responsibility to reset them accordingly.

A sample extending `class ORTModulePreserveState(ORTModule)` could be provided to exemplify the issue and a possible solution

Comment (mindest) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:231]: @baijumeswani Yes [for this model](https://github.com/microsoft/onnxruntime/pull/10705#issue-1154987620) we found that the random state is changed during the shape inference: it is introduced by `sympy.Min()` within [`_infer_Slice`](https://github.com/microsoft/onnxruntime/blob/50a6f095cd2a13e524fba054cf82cfc76fa57185/onnxruntime/python/tools/symbolic_shape_infer.py#L1425).

File: orttraining/orttraining/python/training/ortmodule/_inference_manager.py [hunk 2]
```diff
@@ -86,7 +86,9 @@ def forward(self, *inputs, **kwargs):
             if self._skip_check.is_set(_SkipCheck.SKIP_CHECK_BUILD_GRADIENT) is False or \
                 not self._onnx_models.exported_model:
                 # Exporting module to ONNX for the first time
+                random_states = _utils.get_random_states()
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_inference_manager.py:89]: Since this is common code between training manager and inference manager, may we move it inside `self._export_model`?

Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_inference_manager.py:89]: Since this is common code between training manager and inference manager, may we move it inside `self._export_model`?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 1 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:231]: just for my understanding, are the different ranks performing different operations? If all the ranks are doing the same work, they should all have the same state after export. Maybe I am missing something here.

File: orttraining/orttraining/python/training/ortmodule/_utils.py [hunk 3]
```diff
@@ -25,6 +25,24 @@
...
 import warnings
 from distutils.version import LooseVersion
+import random
+import numpy as np
+
+def get_random_states():
+    r_state = random.getstate()
+    np_state = np.random.get_state()
+    torch_state = torch.get_rng_state()
+    torch_cuda_state = torch.cuda.get_rng_state() if torch.cuda.is_available() else None
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:35]: Could there be other libraries that we would need to set in the future? Is there a way to avoid having to do this? Or maybe let user control this with using `_use_deterministic_algorithms`?

Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:35]: Could there be other libraries that we would need to set in the future? Is there a way to avoid having to do this? Or maybe let user control this with using `_use_deterministic_algorithms`?

**Comment by baijumeswani** [comment]
May we add a unit test for this change?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 1 above)
Comment (mindest) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:231]: @baijumeswani Different ranks should be performing the same operations. But the operation `sympy.Min()` is found to randomly change the random state, e.g. for the following simple case, starting from the same random state, the random state is different after `Min()` anyway. This explains why it results in different random states on different ranks. The root cause is inside those sympy functions, I think it is safer to do the save&restore on ORT side.

```py
import random
from sympy import Min, Symbol

a, b = Symbol('a'), Symbol('b')

random.seed(0)
state_before_0 = random.getstate()
Min(a, b)
state_after_0 = random.getstate()

random.seed(0)
state_before_1 = random.getstate()
Min(a, b)
state_after_1 = random.getstate()

print(state_before_0 == state_before_1) # True
print(state_after_0 == state_after_1) # False
```

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py]: ok, thanks for the explanation.

File: orttraining/orttraining/python/training/ortmodule/_utils.py (hunk 3 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_utils.py:35]: Probably `_use_deterministic_algorithms` has a different purpose than the scope of this pull request. I think it is ok to go with the assumptions you have set out so far.

**Comment by baijumeswani** [comment]
Please try to see if you can add a unit test. Otherwise the changes are looking good.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 4]
```diff
@@ -311,6 +311,8 @@ def _export_model(self, *inputs, **kwargs):
         #       Model is not re-exported when the model parameters change. This can happen when the model is a stateful model,
         #       or the user explicitly changed model parameters after the onnx export.
 
+        random_states = _utils.get_random_states()
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:314]: can you add a quick note here why we keep the states here?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 5]
```diff
@@ -311,6 +311,10 @@ def _export_model(self, *inputs, **kwargs):
         #       Model is not re-exported when the model parameters change. This can happen when the model is a stateful model,
         #       or the user explicitly changed model parameters after the onnx export.
 
+        # record random states here and restore later in case any of them gets changed during the export,
+        # e.g., some sympy functions in symbolic_shape_infer will change Python's random state.
+        random_states = _utils.get_random_states()
```
[the remark this answers is not in the document]
Comment (mindest) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:316]: @thiagocrepaldi sorry I did not make it clear: the random state change actually occurs during the [symbolic shape inference](https://github.com/microsoft/onnxruntime/blob/9f5db78aac10699d0dbce6a59010fb53502d1152/orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py#L330-L332) for this model's case. More precisely, it happens within the `_infer_Slice`. I added it here just in case some future change would also touch any of the random states.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 6]
```diff
@@ -311,6 +311,10 @@ def _export_model(self, *inputs, **kwargs):
         #       Model is not re-exported when the model parameters change. This can happen when the model is a stateful model,
         #       or the user explicitly changed model parameters after the onnx export.
 
+        # record random states here and restore later in case any of them gets changed during the export,
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:314]: 'R'

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 7]
```diff
@@ -4818,3 +4818,38 @@ def run_step(model, x):
...
+
+def test_random_states_unchanged_for_ortmodule():
+    import numpy
+    class NeuralNetSlice(torch.nn.Module):
+        def __init__(self):
+            super(NeuralNetSlice, self).__init__()
+            self.dim = 32
+
+        def forward(self, x):
+            # This slice operation will call sympy.Min() when exporting
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:4831]: Then sympy.Min() internally changed the state, right? Can we make this comment complete, for guys to understand the problem. :)

Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:4831]: Oh, you'd better add following env in case ORTModule runs failed caused the UT to fallback to PyTorch. 

os.environ['ORTMODULE_FALLBACK_POLICY'] = "FALLBACK_DISABLE"
os.environ['ORTMODULE_FALLBACK_RETRY'] = "False"

**Review by pengwa — DISMISSED** [review]
Thanks for catching this problem. Overall LGTM :) Very tricky to debug from the convergence issue.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 7 above)
Comment (mindest) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: Found `ORTMODULE_FALLBACK_POLICY` is already taken care of [here](https://github.com/microsoft/onnxruntime/blob/040c0645e2ce600df1284d314976f797fc5f61c0/orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py#L351). Added `ORTMODULE_FALLBACK_RETRY` only.

**Review by pengwa — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]
Looks good. Thanks for finding and resolving this.
